### PR TITLE
Generic Host 3.0 update

### LINKED ISSUE #11969: Generic Host updates for 3.0
For some of the following tasks it may be appropriate to create separate issues and PRs.

* [x] Host section of Fundamentals overview - review and update as needed. It already has 3.0 content in a moniker block. In the 2.x block add text advising that in 3.0 Generic Host is recommended over Web Host. #12794
* [x] Table of Contents - Reorder the Web Host and Generic Host nodes so Generic Host is first. #12816
* [x] Intro section of Web Host doc - in both 3.0 and 2.x moniker blocks include text advising that in 3.0 Generic Host is recommended over Web Host. #12873
* [x] Other docs that link to the Web Host doc - change links to point to the Generic Host doc or the Host section of the Fundamentals overview as appropriate. #12847
* [ ] App startup doc - injection restrictions: https://github.com/aspnet/Announcements/issues/353
* [ ] Generic Host doc - review and update as needed. It currently has 3.0 content only in the intro in a moniker block.

[EDIT by @guardrex: Just noting that this issue comment was updated to the latest guidance from engineering by @tdykstra.]

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: c5f1eeee-d500-82bf-0d0a-b4e0d5d5c175
* Version Independent ID: 2373a3df-6ea3-50c7-d63d-61c51d572511
* Content: [.NET Generic Host](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/host/generic-host?view=aspnetcore-2.2)
* Content Source: [aspnetcore/fundamentals/host/generic-host.md](https://github.com/aspnet/Docs/blob/master/aspnetcore/fundamentals/host/generic-host.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-fundamentals**
* GitHub Login: @guardrex
* Microsoft Alias: **riande**

### PR DESCRIPTION
**Pull request by tdykstra**
Fixes #11969 
[Internal review URL](https://review.docs.microsoft.com/aspnet/core/fundamentals/host/generic-host?view=aspnetcore-3.0&branch=pr-en-us-13072)
* Update for 3.0.
* Cover both ASP.NET and non-ASP.NET scenarios.
* Avoid advanced scenarios in a Fundamentals article.
* Avoid duplicating material available in other docs.

### CODE REVIEW

File: aspnetcore/fundamentals/host/generic-host.md [hunk 1]
```diff
@@ -1,31 +1,494 @@
...
+* [WebRootPath](#webroot)
+
+## Host configuration
+
+Host configuration is used for the following:
+
+* The properties of the <xref:Microsoft.Extensions.Hosting.IHostEnvironment> implementation.
+* The [URLs](#urls) setting in web apps. The host passes this value to the servers through an `IServerAddresses` API.
+
+Host configuration automatically flows to app configuration ([ConfigureAppConfiguration](#configureappconfiguration).
```
Comment (Rick-Anderson) [inline aspnetcore/fundamentals/host/generic-host.md]: Host configuration automatically flows to app configuration ([ConfigureAppConfiguration](#configureappconfiguration)).

Comment (Rick-Anderson) [reply aspnetcore/fundamentals/host/generic-host.md]: Or this next version might be better

Comment (Rick-Anderson) [reply aspnetcore/fundamentals/host/generic-host.md]: Host configuration automatically flows to [app configuration](#configureappconfiguration).

File: aspnetcore/fundamentals/host/generic-host.md [hunk 2]
```diff
@@ -1,31 +1,494 @@
...
+
+To add host configuration, call <xref:Microsoft.Extensions.Hosting.HostBuilder.ConfigureHostConfiguration*> on `IHostBuilder`. `ConfigureHostConfiguration` can be called multiple times with additive results. The host uses whichever option sets a value last on a given key.
+
+The environment variable provider with prefix `DOTNET_` is included by default. For web apps, the environment variable provider with prefix `ASPNETCORE_` is added. The prefix is removed when the environment variables are read. For example, the environment variable value for `ASPNETCORE_ENVIRONMENT` becomes the host configuration value for the `environment` key.
+
+The following example creates host configuration:
+
+[!code-csharp[](generic-host/samples-snapshot/3.x/Program.cs?name=snippet_HostConfig)]
+
+> [!TIP]
```
Comment (Rick-Anderson) [inline aspnetcore/fundamentals/host/generic-host.md]: Too much bling for me. This also seems TMI for an advanced topic. Perhaps something like my next proposal.

Comment (Rick-Anderson) [inline aspnetcore/fundamentals/host/generic-host.md]: Development environment variables can be set in *.json* files.  For more information, see <xref:fundamentals/environments>.

File: aspnetcore/fundamentals/host/generic-host.md [hunk 3]
```diff
@@ -1,31 +1,494 @@
...
+
+The configuration created by `ConfigureAppConfiguration` is available at [HostBuilderContext.Configuration](xref:Microsoft.Extensions.Hosting.HostBuilderContext.Configuration*) for subsequent operations and as a service from DI.
+
+App configuration automatically receives host configuration provided by `ConfigureHostConfiguration`.
+
+The following example creates app configuration:
+
+[!code-csharp[](generic-host/samples-snapshot/3.x/Program.cs?name=snippet_AppConfig)]
+
+> [!TIP]
```
Comment (Rick-Anderson) [inline aspnetcore/fundamentals/host/generic-host.md]: Not a tip bling fan. The content is fine.

File: aspnetcore/fundamentals/host/generic-host.md [hunk 4]
```diff
@@ -1,31 +1,494 @@
...
+> ```xml
+> <ItemGroup>
+>   <Content Include="**\*.json" Exclude="bin\**\*;obj\**\*" 
+>     CopyToOutputDirectory="PreserveNewest" />
+> </ItemGroup>
+> ```
+
+## Settings for all app types
+
+This section lists host settings that apply to non-HTTP as well as HTTP workloads. By default, environment variables used to configure these settings can have a `DOTNET_` or `ASPNETCORE_` prefix.
```
Comment (Rick-Anderson) [inline aspnetcore/fundamentals/host/generic-host.md]: This section lists host settings that apply to non-HTTP and HTTP workloads. By default, environment variables used to configure these settings can have a `DOTNET_` or `ASPNETCORE_` prefix.

File: aspnetcore/fundamentals/host/generic-host.md (hunk 2 above)
Comment (tdykstra) [reply aspnetcore/fundamentals/host/generic-host.md]: Environments doc is missing some of what's in the current text; I'll condense the text and drop the Tip box.

Comment (Rick-Anderson) [reply aspnetcore/fundamentals/host/generic-host.md]: Can you open an issue on the environment doc? That level of detail doesn't belong in this doc.

Comment (tdykstra) [reply aspnetcore/fundamentals/host/generic-host.md]: Agreed on both counts.

File: aspnetcore/fundamentals/host/generic-host.md [hunk 5]
```diff
@@ -1,31 +1,490 @@
...
+```csharp
+public static IHostBuilder CreateHostBuilder(string[] args) =>
+    Host.CreateDefaultBuilder(args)
+        .ConfigureWebHostDefaults(webBuilder =>
+        {
+            webBuilder.UseStartup<Startup>();
+        });
+```
+
+If the app uses Entity Framework Core, don't change the name of the `CreateHostBuilder` method. The [Entity Framework Core tools](/ef/core/miscellaneous/cli/) expect to find a `CreateHostBuilder` method that configures the host without running the app. For more information, see [Design-time DbContext Creation](/ef/core/miscellaneous/cli/dbcontext-creation).
```
Comment (Tratcher) [inline aspnetcore/fundamentals/host/generic-host.md]:
```suggestion
If the app uses Entity Framework Core, don't change the name or signature of the `CreateHostBuilder` method. The [Entity Framework Core tools](/ef/core/miscellaneous/cli/) expect to find a `CreateHostBuilder` method that configures the host without running the app. For more information, see [Design-time DbContext Creation](/ef/core/miscellaneous/cli/dbcontext-creation).
```

File: aspnetcore/fundamentals/host/generic-host.md [hunk 6]
```diff
@@ -1,31 +1,490 @@
...
+
+## Default builder settings 
+
+The <xref:Microsoft.Extensions.Hosting.Host.CreateDefaultBuilder*> method:
+
+* Sets the content root to the path returned by <xref:System.IO.Directory.GetCurrentDirectory*>.
+* Loads host configuration from environment variables prefixed with "DOTNET_".
```
Comment (Tratcher) [inline aspnetcore/fundamentals/host/generic-host.md]: And command line arguments.

File: aspnetcore/fundamentals/host/generic-host.md [hunk 7]
```diff
@@ -1,31 +1,490 @@
...
+  * *appsettings.{Environment}.json*.
+  * [Secret Manager](xref:security/app-secrets) when the app runs in the `Development` environment.
+  * Environment variables.
+  * Command-line arguments.
+* Adds the following [logging](xref:fundamentals/logging/index) providers:
+  * Console
+  * Debug
+  * EventSource
+  * EventLog (only when running on Windows)
+* Enables [scope validation](xref:fundamentals/dependency-injection#scope-validation) when the environment is Development.
```
Comment (Tratcher) [inline aspnetcore/fundamentals/host/generic-host.md]: and dependency validation
https://github.com/aspnet/Extensions/blob/689e0d7644d5c92c1ce2f33cfc02955fa254d7f3/src/DependencyInjection/DI/src/ServiceProviderOptions.cs#L21-L25

File: aspnetcore/fundamentals/host/generic-host.md [hunk 8]
```diff
@@ -1,31 +1,490 @@
...
+Web apps implement the `IWebHostEnvironment` interface, which inherits `IHostEnvironment` and adds:
+
+* [WebRootPath](#webroot)
+
+## Host configuration
+
+Host configuration is used for the following:
+
+* The properties of the <xref:Microsoft.Extensions.Hosting.IHostEnvironment> implementation.
+* The [URLs](#urls) setting in web apps. The host passes this value to the servers through an `IServerAddresses` API.
```
Comment (Tratcher) [inline aspnetcore/fundamentals/host/generic-host.md]: IServerAddresses is pulled from app configuration for generic host.

File: aspnetcore/fundamentals/host/generic-host.md [hunk 9]
```diff
@@ -1,31 +1,490 @@
...
+* [EnvironmentName](#environmentname)
+* [ContentRootPath](#contentrootpath)
+
+Web apps implement the `IWebHostEnvironment` interface, which inherits `IHostEnvironment` and adds:
+
+* [WebRootPath](#webroot)
+
+## Host configuration
+
+Host configuration is used for the following:
```
Comment (Tratcher) [inline aspnetcore/fundamentals/host/generic-host.md]: It's available from HostBuilderContext.Configuration inside ConfigureAppConfiguration. After ConfigureAppConfiguration HostBuilderContext.Configuration is replaced with the app config.
https://github.com/aspnet/Extensions/blob/36a9bf9d3fdea32b548d7001b27e6b19ae78b094/src/Hosting/Hosting/src/HostBuilder.cs#L169-L190

File: aspnetcore/fundamentals/host/generic-host.md [hunk 10]
```diff
@@ -1,31 +1,490 @@
...
+
+* The properties of the <xref:Microsoft.Extensions.Hosting.IHostEnvironment> implementation.
+* The [URLs](#urls) setting in web apps. The host passes this value to the servers through an `IServerAddresses` API.
+
+Host configuration automatically flows to [app configuration](#app-configuration).
+
+To add host configuration, call <xref:Microsoft.Extensions.Hosting.HostBuilder.ConfigureHostConfiguration*> on `IHostBuilder`. `ConfigureHostConfiguration` can be called multiple times with additive results. The host uses whichever option sets a value last on a given key.
+
+The environment variable provider with prefix `DOTNET_` is included by default. For web apps, the environment variable provider with prefix `ASPNETCORE_` is added. The prefix is removed when the environment variables are read. For example, the environment variable value for `ASPNETCORE_ENVIRONMENT` becomes the host configuration value for the `environment` key.
```
Comment (Tratcher) [inline aspnetcore/fundamentals/host/generic-host.md]:
```suggestion
The environment variable provider with prefix `DOTNET_` and command line args are included by CreateDefaultBuilder. For web apps, the environment variable provider with prefix `ASPNETCORE_` is added. The prefix is removed when the environment variables are read. For example, the environment variable value for `ASPNETCORE_ENVIRONMENT` becomes the host configuration value for the `environment` key.
```

File: aspnetcore/fundamentals/host/generic-host.md [hunk 11]
```diff
@@ -1,31 +1,490 @@
...
+```
+
+### URLs
+
+A semicolon-delimited list of IP addresses or host addresses with ports and protocols that the server should listen on for requests. For example, `http://localhost:123`. Use "\*" to indicate that the server should listen for requests on any IP address or hostname using the specified port and protocol (for example, `http://*:5000`). The protocol (`http://` or `https://`) must be included with each URL. Supported formats vary among servers.
+
+**Key**: urls  
+**Type**: *string*  
+**Default**: `http://localhost:5000` and `https://localhost:5001`
+**Environment variable**: `ASPNETCORE_URLS`
```
Comment (Tratcher) [inline aspnetcore/fundamentals/host/generic-host.md]:
```suggestion
**Environment variable**: `<PREFIX_>URLS`
```

File: aspnetcore/fundamentals/host/generic-host.md [hunk 12]
```diff
@@ -1,31 +1,490 @@
...
+Kestrel has its own endpoint configuration API. For more information, see <xref:fundamentals/servers/kestrel#endpoint-configuration>.
+
+### WebRoot
+
+The relative path to the app's static assets.
+
+**Key**: webroot  
+**Type**: *string*  
+**Default**: *(Content Root)/wwwroot*, if the path exists. If the path doesn't exist, a no-op file provider is used.  
+**Environment variable**: `ASPNETCORE_WEBROOT`
```
Comment (Tratcher) [inline aspnetcore/fundamentals/host/generic-host.md]:
```suggestion
**Environment variable**: `<PREFIX_>WEBROOT`
```

File: aspnetcore/fundamentals/host/generic-host.md [hunk 13]
```diff
@@ -1,31 +1,490 @@
...
+
+<xref:Microsoft.Extensions.Hosting.IHostLifetime.WaitForStartAsync*> is called at the start of `StartAsync`, which waits until it's complete before continuing. This can be used to delay startup until signaled by an external event.
+
+### StopAsync
+
+<xref:Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.StopAsync*> attempts to stop the host within the provided timeout.
+
+### WaitForShutdown
+
+<xref:Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.WaitForShutdown*> blocks the calling thread until shutdown is triggered via Ctrl+C/SIGINT or SIGTERM.
```
Comment (Tratcher) [inline aspnetcore/fundamentals/host/generic-host.md]:
```suggestion
<xref:Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.WaitForShutdown*> blocks the calling thread until shutdown is triggered by the IHostLifetime, such as via Ctrl+C/SIGINT or SIGTERM.
```

File: aspnetcore/fundamentals/host/generic-host.md [hunk 14]
```diff
@@ -1,31 +1,490 @@
...
+
+### WaitForShutdownAsync
+
+<xref:Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.WaitForShutdownAsync*> returns a <xref:System.Threading.Tasks.Task> that completes when shutdown is triggered via the given token and calls <xref:Microsoft.Extensions.Hosting.IHost.StopAsync*>.
+
+### External control
+
+External control of the host can be achieved using methods that can be called externally:
```
Comment (Tratcher) [inline aspnetcore/fundamentals/host/generic-host.md]:
```suggestion
Direct control of the host lifetime can be achieved using methods that can be called externally:
```

File: aspnetcore/fundamentals/host/generic-host.md [hunk 15]
```diff
@@ -16,21 +16,18 @@ This article introduces the .NET Core Generic Host (<xref:Microsoft.Extensions.H
...
 
-When a host starts, it calls `IHostedService.StartAsync` on each implementation of <xref:Microsoft.Extensions.Hosting.IHostedService> that is found in the DI container. In a web app, one of the `IHostedService` implementations is a web service that starts an [HTTP server implementation](xref:fundamentals/index#servers). There can also be other `IhostedService` implementations in the same host for background processes. 
+The main reason for including all of the app's interdependent resources in one object is lifetime management: control over app startup and graceful shutdown.
 
-Apps that don't handle HTTP requests can also benefit from a host's ability to manage functions such as DI, logging, configuration, and app lifetime. Such apps have one or more `IHostedService` implementations and no HTTP server implementation. For more information on using the Generic Host for apps other than ASP.NET Core, see <xref:fundamentals/host/hosted-services>.
+## Web Host is deprecated
```
Comment (Tratcher) [inline aspnetcore/fundamentals/host/generic-host.md]: @glennc for messaging?

**Comment by tdykstra** [comment]
@glennc -- I removed the statement that Web Host is deprecated -- let me know if I should put it back.

**Pull request merged**